### PR TITLE
Implementation of `sssom-schema` as a package

### LINKED ISSUE #256: Migrate to SSSOM 1.0
This is split in two stages

- [ ] SSSOM currently floats just a few PRs before version 1.0.  After #226 we want to migrate sssom-py to that model.
- [x] After all PRs is SSSOM are merged https://github.com/mapping-commons/sssom/pulls, we migrate the remaining stuff

### PR DESCRIPTION
**Pull request by hrshdhgd**
This fixes #256 

 - [x] change back to the commented url in `test_parsers.py` line 35-ish.

### CODE REVIEW

File: sssom/__init__.py [hunk 1]
```diff
@@ -11,8 +11,10 @@
 
     __version__ = importlib_metadata.version(__name__)
 
-from .sssom_datamodel import slots  # noqa:401
-from .sssom_datamodel import Mapping, MappingSet  # noqa:401
+# from .sssom_datamodel import slots  # noqa:401
+# from .sssom_datamodel import Mapping, MappingSet  # noqa:401
+from sssom_schema import Mapping, MappingSet, slots  # noqa:401
```
Comment (matentzn) [inline sssom/__init__.py:16]: ❤️

**Review by matentzn — APPROVED** [review]
EXCELLENT!

**Comment by hrshdhgd** [comment]
@matentzn , I have fixed all errors in test except 1. This is an `rdf` file related error and I can't seem to figure out what is happening. I change [this file](https://github.com/mapping-commons/sssom-py/pull/264/files#diff-b7e7d75d1896040a1cfcb1fbfed8fe527ff2e1249fc01728c2815aab2879923c) and the test failing is test_writers.py specifically [this line](https://github.com/mapping-commons/sssom-py/blob/c5acc28f5dde0e3319547f99ee1bc1feba582166/tests/test_writers.py#L40). I dug into `parse_sssom_rdf` and deeper into `from_sssom_rdf` and [this is where](https://github.com/mapping-commons/sssom-py/blob/c5acc28f5dde0e3319547f99ee1bc1feba582166/sssom/parsers.py#L399) the problem lies.

The process just skips the `for` loop in line 399. Is there anything wrong in my change to the rdf file (`tests/data/basic.sssom.rdf`) ?

File: sssom/parsers.py [hunk 2]
```diff
@@ -589,7 +590,9 @@ def from_obographs(
                                     )
                                     mdict["subject_label"] = label
                                     mdict["predicate_id"] = "oboInOwl:hasDbXref"
-                                    mdict["match_type"] = MATCH_TYPE_UNSPECIFIED
+                                    mdict[
+                                        "mapping_justification"
```
Comment (matentzn) [inline sssom/parsers.py:594]: can we start using constants for this, like in constants.py 

MAPPING_JUSTIFICATION="mapping_justification" and then

mdict[MAPPING_JUSTIFICATION]=MAPPING_JUSTFN_UNSPECIFIED

Comment (matentzn) [reply sssom/parsers.py:594]: you also do this in util.py, so you can just move all the constants in there to constants.py and then import them in all other files?

File: tests/data/basic-meta-external.tsv [hunk 3]
```diff
@@ -1,142 +1,142 @@
...
+subject_id	subject_label	predicate_id	predicate_modifier	object_id	object_label	mapping_justification	subject_source	object_source	mapping_tool	confidence	subject_match_field	object_match_field	subject_category	object_category	match_string	comment
+x:appendage	appendage	owl:equivalentClass		y:appendage	appendages	semapv:SemanticSimilarityThresholdMatching	x	y	rdf_matcher	0.840714406	Label	Label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	appendag	.
+x:appendage	appendage	owl:equivalentClass		z:appendage	APPENDAGE	semapv:SemanticSimilarityThresholdMatching	x	z	rdf_matcher	0.840714406	Label	Label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	appendag	.
+x:appendage	appendage	owl:equivalentClass		z:appendage	APPENDAGE	ssemapv:LexicalMatching	x	z	rdf_matcher	0.881856236	Label	Label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	appendage	.
+x:bone_element	bone element	owl:equivalentClass		y:bone	bones	ssemapv:LexicalMatching	x	y	rdf_matcher	0.738796125	oio:hasDbXref	oio:hasDbXref	biolink:AnatomicalEntity	biolink:AnatomicalEntity	UBERON:0001474	.
```
Comment (matentzn) [inline tests/data/basic-meta-external.tsv:5]: `semapv`, not `ssemapv`

Comment (matentzn) [reply tests/data/basic-meta-external.tsv:5]: this happens a few times.

File: tests/data/basic.json [hunk 4]
```diff
@@ -1,2637 +1,2295 @@
...
+      {
+         "subject_id":"x:appendage",
+         "predicate_id":"owl:equivalentClass",
+         "object_id":"y:appendage",
+         "mapping_justification":"semapv:ManualMappingCuration",
+         "subject_label":"appendage",
+         "subject_category":"biolink:AnatomicalEntity",
+         "object_label":"appendages",
+         "object_category":"biolink:AnatomicalEntity",
+         "subject_source":"x",
```
Comment (matentzn) [inline tests/data/basic.json:14]: this should no longer be allowed: https://github.com/mapping-commons/sssom/blob/master/src/sssom_schema/schema/sssom_schema.yaml#L322

File: tests/data/basic.sssom.rdf [hunk 5]
```diff
@@ -142,7 +142,7 @@ x:bone_element owl:equivalentClass y:bone,
             sssom:confidence "0.8407144059077161"^^<xsd:double> ;
             sssom:mapping_tool "rdf_matcher" ;
             sssom:match_string "appendag" ;
-            sssom:match_type "Lexical|Unspecified" ;
+            sssom:mapping_justification "semapv:UnspecifiedMatching" ;
```
Comment (matentzn) [inline tests/data/basic.sssom.rdf:145]: this should be:

```
sssom:mapping_justification semapv:UnspecifiedMatching
```

not

```
sssom:mapping_justification "semapv:UnspecifiedMatching"
```

(quotes). See https://github.com/mapping-commons/sssom/blob/master/src/sssom_schema/schema/sssom_schema.yaml#L213

File: tests/data/basic.sssom.rdf [hunk 6]
```diff
@@ -363,7 +363,7 @@ x:bone_element owl:equivalentClass y:bone,
             sssom:confidence "0.2612038749637414"^^<xsd:double> ;
             sssom:mapping_tool "rdf_matcher" ;
             sssom:match_string "bone" ;
-            sssom:match_type "Lexical" ;
+            sssom:mapping_justification "ssemapv:LexicalMatching" ;
```
Comment (matentzn) [inline tests/data/basic.sssom.rdf:366]: again the `ssemapv` instead of `semapv`.

File: tests/test_merge.py [hunk 7]
```diff
@@ -21,7 +21,7 @@ def setUp(self) -> None:
     def test_merge_multiple_inputs(self):
         """Test merging of multiple msdfs."""
         merged_msdf = merge_msdf(*self.msdfs)
-        self.assertEqual(71, len(merged_msdf.df))
+        self.assertEqual(20, len(merged_msdf.df))
```
Comment (matentzn) [inline tests/test_merge.py:24]: Why this change?

File: tox.ini [hunk 8]
```diff
@@ -14,8 +14,8 @@ envlist =
 [testenv]
 commands =
     pytest
-extras =
-    test
+; extras =
+;     test
```
Comment (matentzn) [inline tox.ini:18]: why this change?

File: tests/test_reconcile.py [hunk 9]
```diff
@@ -31,7 +31,7 @@ def test_merge(self):
         msdf2 = parse_sssom_table(data_dir / "basic2.tsv")
         merged_msdf = merge_msdf(msdf1, msdf2)
 
-        self.assertEqual(71, len(merged_msdf.df))
+        self.assertEqual(24, len(merged_msdf.df))
```
Comment (matentzn) [inline tests/test_reconcile.py:34]: Why this change?

File: tests/test_parsers.py [hunk 10]
```diff
@@ -173,17 +174,20 @@ def test_parse_sssom_json(self):
             f"{self.json_file} has the wrong number of mappings.",
         )
 
-    def test_piped_element_to_list(self):
```
Comment (matentzn) [inline tests/test_parsers.py:176]: We should still have a test for this just on a different element (one that has "multivalued: true".

File: tests/test_parsers.py [hunk 11]
```diff
@@ -32,7 +31,9 @@ class TestParse(unittest.TestCase):
 
     def setUp(self) -> None:
         """Set up the test case."""
-        self.df_url = "https://raw.githubusercontent.com/mapping-commons/sssom-py/master/tests/data/basic.tsv"
+        # TODO: change back to the commented url.
```
Comment (matentzn) [inline tests/test_parsers.py:34]: Add this as a checkbox to the PR description to really not forget it :D

File: tests/data/basic5.tsv [hunk 12]
```diff
@@ -13,57 +13,57 @@
...
-c1:something	YYYYY	owl:equivalentClass		d1:something	yyyyyy	Lexical	c	d	rdf_matcher	0.81	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	xxxxx	mock data
-d1:something	YYYYY	owl:equivalentClass	Not	a1:something	yyyyyy	Lexical	d	a	rdf_matcher	0.82	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	xxxxx	mock data
-a1:something	XYXYX	owl:equivalentClass		c1:something	xyxyxy	Lexical	a	c	rdf_matcher	0.83	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	xxxxx	mock data
-c1:something	YXYXY	owl:equivalentClass		b1:something	yxyxyx	Lexical	c	b	rdf_matcher	0.84	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	xxxxx	mock data
\ No newline at end of file
+subject_id	subject_label	predicate_id	predicate_modifier	object_id	object_label	mapping_justification	subject_source	object_source	mapping_tool	confidence	subject_match_field	object_match_field	subject_category	object_category	match_string	comment
+x1:appendage	appendage	owl:equivalentClass		y1:appendage	appendages	ssemapv:LexicalMatching	x	z	rdf_matcher	0.840714406	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	organ	.
```
Comment (matentzn) [inline tests/data/basic5.tsv:17]: ssemapv instead of semapv

File: tests/data/basic4.tsv [hunk 13]
```diff
@@ -13,57 +13,57 @@
...
+x2:liver	liver	owl:equivalentClass		z2:organ	ORGAN	ssemapv:LexicalMatching	x	z	rdf_matcher	0.738796125	oio:hasDbXref	oio:hasDbXref	biolink:AnatomicalEntity	biolink:AnatomicalEntity	BAd2:ORGAN	.
+x2:lung	lung	owl:equivalentClass		y2:lung	lungs	ssemapv:LexicalMatching	x	z	rdf_matcher	0.881856236	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	hindlimb	.
+x2:lung	lung	owl:equivalentClass		z2:organ	ORGAN	semapv:SemanticSimilarityThresholdMatching	x	z	rdf_matcher	0.840714406	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	hindlimb	.
+x2:organ	organ	owl:equivalentClass		y2:organ	organs	ssemapv:LexicalMatching	x	y	rdf_matcher	0.738796125	oio:hasDbXref	oio:hasDbXref	biolink:AnatomicalEntity	biolink:AnatomicalEntity	UBERON:0002107	.
+x2:organ	organ	owl:equivalentClass		z2:organ	ORGAN	semapv:SemanticSimilarityThresholdMatching	x	y	rdf_matcher	0.840714406	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	liver	.
+x2:region	region	owl:equivalentClass		y2:region	regions	ssemapv:LexicalMatching	x	z	rdf_matcher	0.738796125	oio:hasDbXref	oio:hasDbXref	biolink:AnatomicalEntity	biolink:AnatomicalEntity	BAd2:ORGAN	.
```
Comment (matentzn) [inline tests/data/basic4.tsv:33]: ssemapv instead of semapv

File: tests/data/basic2.tsv [hunk 14]
```diff
@@ -13,130 +13,130 @@
...
+subject_id	subject_label	predicate_id	predicate_modifier	object_id	object_label	mapping_justification	subject_source	object_source	mapping_tool	confidence	subject_match_field	object_match_field	subject_category	object_category	match_string	comment
+x:FOO	FOO	owl:equivalentClass		y:FOO	FOO	semapv:SemanticSimilarityThresholdMatching	x	y	rdf_matcher	0.840714406	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	foo	.
+x:appendage	appendage	owl:equivalentClass		y:appendage	appendages	semapv:ManualMappingCuration	x	y	rdf_matcher	0.840714406	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	appendag	.
+x:appendage	appendage	owl:equivalentClass		z:appendage	APPENDAGE	semapv:ManualMappingCuration	x	z	rdf_matcher	0.840714406	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	appendag	.
+x:appendage	appendage	owl:equivalentClass		z:appendage	APPENDAGE	ssemapv:LexicalMatching	x	z	rdf_matcher	0.881856236	rdfs:label	rdfs:label	biolink:AnatomicalEntity	biolink:AnatomicalEntity	appendage	.
```
Comment (matentzn) [inline tests/data/basic2.tsv:20]: ssemapv instead of semapv

**Review by matentzn — COMMENTED** [review]
Great progress 🚀

File: sssom/internal_context.py [hunk 15]
```diff
@@ -2,7 +2,7 @@
 
 # This is autogenerated and super hacky.
 multivalued_slots = [
-    "match_type",
+    "mapping_justification",
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply sssom/internal_context.py]: Done

File: tox.ini (hunk 8 above)
Comment (hrshdhgd) [reply tox.ini:18]: I was trying something .... I had a suspicion it was causing some error. It is not and hence uncommented again.

File: tests/test_merge.py (hunk 7 above)
Comment (hrshdhgd) [reply tests/test_merge.py:24]: with `match_type` changing to be single valued, duplicate rows get filtered out. So length o dataframes will change.

File: tests/test_reconcile.py (hunk 9 above)
Comment (hrshdhgd) [reply tests/test_reconcile.py:34]: with `match_type` changing to be single valued, duplicate rows get filtered out. So length o dataframes will change.

File: sssom/constants.py [hunk 16]
```diff
@@ -1,15 +1,21 @@
...
 
+import pkg_resources
 from linkml_runtime.utils.schema_as_dict import schema_as_dict
 from linkml_runtime.utils.schemaview import SchemaView
 
+# from linkml_runtime.utils.introspection import package_schemaview
+
 HERE = pathlib.Path(__file__).parent.resolve()
-SCHEMA_YAML = os.path.join(HERE, "sssom.yaml")
+# SCHEMA_YAML = os.path.join(HERE, "sssom.yaml")
```
Comment (matentzn) [inline sssom/constants.py:12]: why comment, could be deleted?

**Review by matentzn — COMMENTED** [review]
Super sweet changes man! Amazing!

File: sssom/util.py [hunk 17]
```diff
@@ -605,7 +608,7 @@ def deal_with_negation(df: pd.DataFrame) -> pd.DataFrame:
         PREDICATE_ID,
         OBJECT_ID,
         CONFIDENCE,
-        MATCH_TYPE,
+        MAPPING_JUSTFCN,
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply sssom/util.py]: Done ...makes sense . The initial intent was not making a line of code too long, but `black` takes care of it.

File: tests/data/basic5.tsv (hunk 12 above)
Comment (hrshdhgd) [reply tests/data/basic5.tsv]: Fixed.

File: tests/data/basic4.tsv (hunk 13 above)
Comment (hrshdhgd) [reply tests/data/basic4.tsv]: fixed

File: tests/data/basic2.tsv (hunk 14 above)
Comment (hrshdhgd) [reply tests/data/basic2.tsv]: fixed

File: sssom/constants.py (hunk 16 above)
Comment (hrshdhgd) [reply sssom/constants.py]: Done

File: sssom/internal_context.py [hunk 18]
```diff
@@ -1,23 +1,7 @@
...
-    "reviewer_label",
-    "license",
-    "mapping_set_source",
-    "subject_match_field",
-    "object_match_field",
-    "match_string",
-    "subject_preprocessing",
-    "object_preprocessing",
-    "see_also",
-]
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply sssom/internal_context.py:20]: I'm glad you mentioned this. I was wondering what generated this. Fixed!

File: tox.ini [hunk 19]
```diff
@@ -12,6 +12,7 @@ envlist =
     py
 
 [testenv]
+whitelist_externals = pytest
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply tox.ini:15]: Uses `pytest` in the virtual env rather than installing within the `tox` env again. This takes care of an error `tox` was throwing.

File: tests/data/basic3.tsv [hunk 20]
```diff
@@ -3,14 +3,15 @@
 #mapping_tool: "https://github.com/cmungall/rdf_matcher"
 #creator_id: "cjm"
 #mapping_date: "2020-05-30"
+#mapping_justification: "abcd"
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply tests/data/basic3.tsv:6]: Good catch, I was testing and forgot to remove.

File: tests/data/basic.json (hunk 4 above)
Comment (hrshdhgd) [reply tests/data/basic.json:14]: Do not follow what you mean.

**Comment by hrshdhgd** [comment]
Hey @matentzn , I've addressed all but one change requests. But I still have my question from my previous comment unanswered.

File: tests/data/basic.json (hunk 4 above)
Comment (matentzn) [reply tests/data/basic.json:14]: subject_source is of type "EntityReference" - this here is a string ("x"), while it should be something like x:1.  This should not parse as a valid uri or curie!

File: sssom/util.py [hunk 21]
```diff
@@ -636,7 +638,10 @@ def deal_with_negation(df: pd.DataFrame) -> pd.DataFrame:
                 (combined_normalized_subset[SUBJECT_ID] == row_1[SUBJECT_ID])
                 & (combined_normalized_subset[OBJECT_ID] == row_1[OBJECT_ID])
                 & (combined_normalized_subset[CONFIDENCE] == row_1[CONFIDENCE])
-                & (combined_normalized_subset[MATCH_TYPE] == HUMAN_CURATED_MATCH_TYPE)
+                & (
+                    combined_normalized_subset[MAPPING_JUSTIFICATION]
+                    == "semapv:ManualMappingCuration"
```
Comment (matentzn) [inline sssom/util.py:643]: Better use a token here again rather than a string, maybe something like `SEMPAV.ManualMappingCuration` in the SEMAPV enum in constants?

File: tests/data/basic.sssom.rdf (hunk 5 above)
Comment (matentzn) [reply tests/data/basic.sssom.rdf:145]: This issue here still needs to be resolved.

File: tests/data/basic.json (hunk 4 above)
Comment (matentzn) [reply tests/data/basic.json:14]: I dont understand how it can pass validation! It really should not.

**Comment by matentzn** [comment]
> The process just skips the for loop in line 399. Is there anything wrong in my change to the rdf file (tests/data/basic.sssom.rdf) ?

search the code for _all_ occurrences of `http://w3id.org/sssom/` and replace them by `https://w3id.org/sssom/` - hopefully this fixes your RDF issue. We recently had to change the base IRI of sssom  cause it was wrong!

**Comment by hrshdhgd** [comment]
@matentzn , I'm looking at the requirements:
```
    scikit_learn
    scipy
```

Does `sssom-py` depend on these? Could you please double check the requirements if it has more listed than what we need?

**Comment by hrshdhgd** [comment]
> > The process just skips the for loop in line 399. Is there anything wrong in my change to the rdf file (tests/data/basic.sssom.rdf) ?
> 
> search the code for _all_ occurrences of `http://w3id.org/sssom/` and replace them by `https://w3id.org/sssom/` - hopefully this fixes your RDF issue. We recently had to change the base IRI of sssom cause it was wrong!

Yup,  that fixed it. Now there's a new issue I do not understand. 
```
ERROR: Could not find a version that satisfies the requirement sssom-schema (from versions: none)
```
It worked all this time and now all of a sudden this. I cannot recreate this locally.

**Comment by matentzn** [comment]
1. correlation command depends on scipy: `from scipy.stats import chi2_contingency`
2. sckit learn seems unnecessary (drop)
3. rest seems necessary.

File: tests/data/basic.json (hunk 4 above)
Comment (matentzn) [reply tests/data/basic.json:14]: see https://github.com/mapping-commons/sssom-py/issues/265

File: tests/data/basic.json [hunk 22]
```diff
@@ -1,2637 +1,2295 @@
...
+         "object_source":"y",
```
[the remark this answers is not in the document]
Comment (matentzn) [reply tests/data/basic.json:15]: https://github.com/mapping-commons/sssom-py/issues/265

File: sssom/writers.py [hunk 23]
```diff
@@ -414,7 +416,7 @@ def to_fhir_json(msdf: MappingSetDataFrame) -> Dict:
                                         "url": "http://example.org/fhir/StructureDefinition/mapping_justification",
                                         "ValueString": row.get(
                                             "mapping_justification",
-                                            row.get("match_type", ""),
+                                            row.get("mapping_justification", ""),
```
[the remark this answers is not in the document]
Comment (matentzn) [reply sssom/writers.py:419]: https://github.com/mapping-commons/sssom-py/issues/267

**Review by matentzn — APPROVED** [review]

**Review by matentzn — APPROVED** [review]
